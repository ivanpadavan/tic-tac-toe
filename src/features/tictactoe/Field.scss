.field {
  overflow: hidden;
  grid-area: field;
}
.container {
  width: 100%;
  height: 100%;
}
.cell {
  position: absolute;
  width: 50px;
  height: 50px;
  background-color: white;
  display: flex;
  justify-content: center;
  align-items: center;
  border: 1px solid black;
}
