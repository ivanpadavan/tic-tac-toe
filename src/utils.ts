export const sum = (a: number, b: number) => a + b;
export const range = (n: number): Array<number> =>
  Array.from(Array(n), (_, idx) => idx);
