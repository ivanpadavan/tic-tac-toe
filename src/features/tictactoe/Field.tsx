import React, { ComponentType, useMemo, useRef, useState } from 'react';
import { useElementSize, usePrevious } from '../../app/hooks';
import { range } from '../../utils';
import Cell from './Cell';
import './Field.scss';
import { Coordinate } from './types';

interface FieldProps {
  overflowCells: number;
  xOffset: number;
  yOffset: number;
  cellSize: number;
}
interface InfiniteFieldProps {
  cmp: ComponentType<Coordinate>;
  throttle: number;
}

function arrayOfIndices(
  size: number,
  props: Pick<FieldProps, 'cellSize' | 'overflowCells'>
): [number, number[]] {
  let cellsOnScreen =
    Math.ceil(size / props.cellSize) + props.overflowCells * 2;
  cellsOnScreen = cellsOnScreen % 2 ? cellsOnScreen : cellsOnScreen + 1;
  const brunch = Math.floor(cellsOnScreen / 2);
  return [
    brunch * props.cellSize + size / 2 - (cellsOnScreen * props.cellSize) / 2,
    range(cellsOnScreen).map((_, it) => it - brunch),
  ];
}
function appendOffset(arr: number[], cellSize: number, offset: number) {
  return arr.map((v) => v - Math.ceil(offset / cellSize));
}
export const InfiniteField = (hocProps: InfiniteFieldProps) =>
  function Cmp(props: FieldProps) {
    const { cellSize, overflowCells, xOffset, yOffset } = props;

    const field = useRef<HTMLDivElement>(null);
    const [transition, setTransition] = useState('');
    const fieldSize = useElementSize(hocProps.throttle, field.current);

    const prevFieldSize = usePrevious(fieldSize);
    if (transition !== '' && fieldSize !== prevFieldSize) {
      console.log(fieldSize, prevFieldSize);
      setTransition('');
    }

    const [[translateX, arrX], [translateY, arrY]] = useMemo(
      () => [
        arrayOfIndices(fieldSize.width ?? 0, { cellSize, overflowCells }),
        arrayOfIndices(fieldSize.height ?? 0, { cellSize, overflowCells }),
      ],
      [fieldSize, cellSize, overflowCells]
    );
    const [translate, setTranslate] = useState<Coordinate>({ x: 0, y: 0 });
    if (
      field.current &&
      (translate.x !== translateX || translate.y !== translateY)
    ) {
      setTranslate({ x: translateX, y: translateY });
    }
    const computedCells = useMemo(() => {
      return appendOffset(arrX, cellSize, xOffset).flatMap((x) =>
        appendOffset(arrY, cellSize, yOffset).map((y) => ({ x, y }))
      );
    }, [arrX, arrY, xOffset, yOffset, cellSize]);
    const [cells, setCells] = useState<Coordinate[]>([]);
    if (cells !== computedCells) setCells(computedCells);
    // set transition after initialization
    if (translate.x !== 0)
      setTimeout(() => setTransition('transform .3s linear'), 1e3);

    return (
      <div className='field' ref={field}>
        <div
          className='container'
          style={{
            transform: `translate(${translate.x + props.xOffset}px, ${
              translate.y + props.yOffset
            }px)`,
            transition,
          }}>
          {cells.map(({ x, y }) => (
            <hocProps.cmp
              key={x.toString() + '_' + y.toString()}
              x={x}
              y={y}></hocProps.cmp>
          ))}
        </div>
      </div>
    );
  };
const Field = InfiniteField({ cmp: Cell, throttle: 100 });
export default Field;
