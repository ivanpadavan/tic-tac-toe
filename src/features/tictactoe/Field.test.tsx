import { fireEvent, render } from '@testing-library/react';
import { range, sum } from '../../utils';
import { InfiniteField } from './Field';

const TestField = InfiniteField({
  throttle: 0,
  cmp: function SimpleCell(props) {
    return (
      <div>
        {props.x} {props.y}
      </div>
    );
  },
});

it('renders at least one cell and overflowCells describes additional circles', () => {
  const numOfCells = (overflowCells: number) => {
    return range(overflowCells + 1)
      .map((v) => {
        if (v === 0) return 1;
        const side = v * 2 + 1;
        return side * 4 - 4;
      })
      .reduce(sum);
  };
  for (let overflowCells = 0; overflowCells < 5; overflowCells++) {
    const { container } = render(
      <TestField
        overflowCells={overflowCells}
        xOffset={0}
        yOffset={0}
        cellSize={50}
      />
    );
    const received = container.querySelector('.container')!.children.length;
    const expected = numOfCells(overflowCells);
    expect(received).toEqual(expected);
  }
});
it('rerenders when resizes', () => {
  const { container } = render(
    <TestField overflowCells={0} xOffset={0} yOffset={0} cellSize={1} />
  );
  let side = 0;
  const square = () => ({ width: side, height: side });
  const numOfCells = () =>
    container.querySelector('.container')!.children.length;
  container.querySelector('.field')!.getBoundingClientRect = () =>
    square() as DOMRect;
  const results = [];
  while (side <= 10) {
    side++;
    fireEvent(window, new Event('resize'));
    results.push({ numOfCells: numOfCells(), side });
  }
  expect(results).toMatchSnapshot();
});
it('handles offsets', () => {
  const testOffsets = range(3).map((v) => v - 1);
  const results = testOffsets.flatMap((xOffset) =>
    testOffsets.map((yOffset) => {
      const { container } = render(
        <TestField
          overflowCells={1}
          xOffset={xOffset}
          yOffset={yOffset}
          cellSize={1}
        />
      );
      const nodeTexts = [
        ...container.querySelector('.container')!.children,
      ].map((it) => it.textContent);
      const rowLen = Math.sqrt(nodeTexts.length);
      const key = { xOffset, yOffset };
      const value =
        '\n' +
        range(rowLen)
          .map(
            () =>
              '(' +
              range(rowLen)
                .map(() => nodeTexts.shift())
                .join(') (') +
              ')'
          )
          .join('\n');
      return [key, value] as const;
    })
  );
  expect(new Map(results)).toMatchSnapshot();
});
