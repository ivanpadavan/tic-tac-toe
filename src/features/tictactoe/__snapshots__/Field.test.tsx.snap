// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles offsets 1`] = `
Map {
  Object {
    "xOffset": -1,
    "yOffset": -1,
  } => "
(0 0) (0 1) (0 2)
(1 0) (1 1) (1 2)
(2 0) (2 1) (2 2)",
  Object {
    "xOffset": -1,
    "yOffset": 0,
  } => "
(0 -1) (0 0) (0 1)
(1 -1) (1 0) (1 1)
(2 -1) (2 0) (2 1)",
  Object {
    "xOffset": -1,
    "yOffset": 1,
  } => "
(0 -2) (0 -1) (0 0)
(1 -2) (1 -1) (1 0)
(2 -2) (2 -1) (2 0)",
  Object {
    "xOffset": 0,
    "yOffset": -1,
  } => "
(-1 0) (-1 1) (-1 2)
(0 0) (0 1) (0 2)
(1 0) (1 1) (1 2)",
  Object {
    "xOffset": 0,
    "yOffset": 0,
  } => "
(-1 -1) (-1 0) (-1 1)
(0 -1) (0 0) (0 1)
(1 -1) (1 0) (1 1)",
  Object {
    "xOffset": 0,
    "yOffset": 1,
  } => "
(-1 -2) (-1 -1) (-1 0)
(0 -2) (0 -1) (0 0)
(1 -2) (1 -1) (1 0)",
  Object {
    "xOffset": 1,
    "yOffset": -1,
  } => "
(-2 0) (-2 1) (-2 2)
(-1 0) (-1 1) (-1 2)
(0 0) (0 1) (0 2)",
  Object {
    "xOffset": 1,
    "yOffset": 0,
  } => "
(-2 -1) (-2 0) (-2 1)
(-1 -1) (-1 0) (-1 1)
(0 -1) (0 0) (0 1)",
  Object {
    "xOffset": 1,
    "yOffset": 1,
  } => "
(-2 -2) (-2 -1) (-2 0)
(-1 -2) (-1 -1) (-1 0)
(0 -2) (0 -1) (0 0)",
}
`;

exports[`rerenders when resizes 1`] = `
Array [
  Object {
    "numOfCells": 1,
    "side": 1,
  },
  Object {
    "numOfCells": 9,
    "side": 2,
  },
  Object {
    "numOfCells": 9,
    "side": 3,
  },
  Object {
    "numOfCells": 25,
    "side": 4,
  },
  Object {
    "numOfCells": 25,
    "side": 5,
  },
  Object {
    "numOfCells": 49,
    "side": 6,
  },
  Object {
    "numOfCells": 49,
    "side": 7,
  },
  Object {
    "numOfCells": 81,
    "side": 8,
  },
  Object {
    "numOfCells": 81,
    "side": 9,
  },
  Object {
    "numOfCells": 121,
    "side": 10,
  },
  Object {
    "numOfCells": 121,
    "side": 11,
  },
]
`;
